### PR TITLE
PSR-18: Separate ConnectException from RequestException

### PR DESCRIPTION
**Pull request by davidgrayston**
Guzzle version affected: `7.0.0-beta.1`

_Relates to #2507 / https://github.com/php-http/httplug/issues/155_

### Description
Update exception hierarchy to conform with https://www.php-fig.org/psr/psr-18/meta/
>The domain exceptions NetworkExceptionInterface and RequestExceptionInterface define a contract very similar to each other. The chosen approach is to not let them extend each other because inheritance does not make sense in the domain model. A RequestExceptionInterface is simply not a NetworkExceptionInterface.

### Changed
- `ConnectException` now extends `TransferException` instead of `RequestException`
  > Note: request, response and handler context methods have been moved into traits for reuse - there might be a neater/preferable way

#### New hierarchy:
```
\RuntimeException
-- TransferException
-- -- ConnectException (Psr\Http\Client\NetworkExceptionInterface)
-- -- RequestException (Psr\Http\Client\RequestExceptionInterface)
-- -- -- TooManyRedirectsException
-- -- -- BadResponseException
-- -- -- -- ClientException
-- -- -- -- ServerException
```

I've tried to do this with as few breaking changes as possible - you could instead introduce a new exception that implements `Psr\Http\Client\RequestExceptionInterface` which would extend `RequestException` like this _(which would be non-breaking)_:
```
\RuntimeException
-- TransferException
-- -- RequestException
-- -- -- ConnectException (Psr\Http\Client\NetworkExceptionInterface)
-- -- -- RequestFailedException (Psr\Http\Client\RequestExceptionInterface)
-- -- -- -- TooManyRedirectsException
-- -- -- -- BadResponseException
-- -- -- -- -- ClientException
-- -- -- -- -- ServerException
```

### CODE REVIEW

File: src/Exception/ConnectException.php [hunk 1]
```diff
@@ -10,22 +12,32 @@
...
         array $handlerContext = []
     ) {
-        parent::__construct($message, $request, null, $previous, $handlerContext);
+        parent::__construct($message, 0, $previous);
+        $this->setRequest($request);
+        $this->setHandlerContext($handlerContext);
     }
 
+    /**
+     * @deprecated 8.0.0
```
Comment (GrahamCampbell) [inline src/Exception/ConnectException.php]: We can't mark this as deprecated until the 7.1.0 release, so these annotations will need to come in a followup PR after this one, if this is merged into what will be 7.0.0.

File: src/Exception/ConnectException.php [hunk 2]
```diff
@@ -10,15 +12,19 @@
  *
  * Note that no response is present for a ConnectException
  */
-class ConnectException extends RequestException implements NetworkExceptionInterface
+class ConnectException extends TransferException implements NetworkExceptionInterface
```
Comment (gmponos) [inline src/Exception/ConnectException.php]: If this gets approved we will need to update the UPGRADE guide and also the changelog.. Added as a note here in order not to forget.

Comment (davidgrayston) [reply src/Exception/ConnectException.php]: I've added a note to _UPGRADING.md_ in 3edf0a8

**Comment by davidgrayston** [comment]
@gmponos I can remove use of traits if preferable?

I was also wondering if the following methods could be removed with a note added to `UPGRADING.md`?
- `GuzzleHttp\Exception\ConnectException::getResponse()`
- `GuzzleHttp\Exception\ConnectException::hasResponse()`

File: src/Exception/RequestException.php [hunk 3]
```diff
@@ -11,15 +13,11 @@
  */
 class RequestException extends TransferException implements RequestExceptionInterface
 {
-    /** @var RequestInterface */
-    private $request;
+    use RequestAwareTrait, HandlerContextAwareTrait;
```
Comment (GrahamCampbell) [inline src/Exception/RequestException.php]: If we are following PSR-12 and/or Symfony code style guidelines, then this needs to be:

```php
use RequestAwareTrait;
use HandlerContextAwareTrait;
```

Comment (davidgrayston) [reply src/Exception/RequestException.php]: Fixed in 8ad0f9d

**Comment by davidgrayston** [comment]
Would you prefer the PHPDoc for `getHandlerContext()` be duplicated in each exception, or could this be moved into a handler context interface, with each implementation using inheritdoc?

**Comment by Nyholm** [comment]
You can duplicate it. 
It is "just exceptions" and it is just twice.

**Comment by Nyholm** [comment]
Generally, change log updates should not be done in a PR. That will lead to merge conflicts.

**Comment by Tobion** [comment]
With these changes, we should also change RequestException to make the response non-optional. The response was only optional for ConnectException. But now ConnectException does not extend RequestException anymore. All sub-classes of RequestException have a response now AFAIK.

**Comment by Tobion** [comment]
Looks like TooManyRedirectsException does not have a response but should have. The response is just not forwarded from what I see.

**Comment by Nyholm** [comment]
Thank you @Tobion

**Comment by sergiz** [comment]
The docs haven't been updated and still display old hiearchy: https://docs.guzzlephp.org/en/stable/quickstart.html#exceptions

**Pull request merged**